### PR TITLE
[ROCm] add quant op and port rccl

### PR DESCRIPTION
**Pull request by feifei14119**

### CODE REVIEW

File: src/fastertransformer/core/Buffer.h [hunk 1]
```diff
@@ -66,7 +66,7 @@ class Buffer {
     inline T* data() const {
         static_assert(!std::is_same<T, void>::value);
         FT_CHECK_WITH_INFO(
-            type_ == getTensorType<T>(),
+            type_ == getTensorType<T>() || type_ == DataType::TYPE_INT4X2,
```
Comment (netaddi) [inline src/fastertransformer/core/Buffer.h:69]: Don't change this, to get data pointer for quantized buffer, please use non-template version of data() and cast manually.

File: src/fastertransformer/core/Buffer.h [hunk 2]
```diff
@@ -117,18 +117,38 @@ class Buffer {
 
     template<typename T>
     std::string debugDataString(size_t count) const {
+        if(where_ == MemoryType::MEMORY_GPU)
+            return "Device buffer data can NOT be dump, please use Device->clone() to convert to cpu buffer.";
```
Comment (netaddi) [inline src/fastertransformer/core/Buffer.h:121]: please add brackets for sentences under if condition even if there's only one line.

File: src/fastertransformer/core/Buffer.h [hunk 3]
```diff
@@ -117,18 +117,38 @@ class Buffer {
...
+
         auto base = data<T>();
         auto total_size = size();
         std::ostringstream oss;
         auto data_size = std::min(count, total_size);
-        for (size_t i = 0; i < data_size; i++) {
-            oss << base[i] << ", ";
+        if(type_ == DataType::TYPE_QINT4X2 || type_ == DataType::TYPE_INT4X2)
+        {
```
Comment (netaddi) [inline src/fastertransformer/core/Buffer.h:128]: no new line before the left brace.

**Review by netaddi — CHANGES_REQUESTED** [review]

File: src/fastertransformer/core/Buffer.h (hunk 3 above)
Comment (feifei14119) [reply src/fastertransformer/core/Buffer.h]: Fixed. Thanks~

File: src/fastertransformer/core/Buffer.h (hunk 1 above)
Comment (feifei14119) [reply src/fastertransformer/core/Buffer.h]: Fixed. Thanks~

File: src/fastertransformer/core/Buffer.h (hunk 2 above)
Comment (feifei14119) [reply src/fastertransformer/core/Buffer.h:121]: Fixed. Thanks~

**Review by netaddi — APPROVED** [review]

**Pull request merged**